extensions/librdkafka/ConsumeKafka.cpp

```diff
@@ -0,0 +1,553 @@
+}
+
+std::string ConsumeKafka::resolve_duplicate_headers(const std::vector<std::string>& matching_headers) const {
+  if (MSG_HEADER_KEEP_FIRST == duplicate_header_handling_) {
+    return matching_headers.front();
+  }
+  if (MSG_HEADER_KEEP_LATEST == duplicate_header_handling_) {
+    return matching_headers.back();
+  }
+  if (MSG_HEADER_COMMA_SEPARATED_MERGE == duplicate_header_handling_) {
+    return utils::StringUtils::join(", ", matching_headers);
+  }
+  throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "\"Duplicate Header Handling\" property not recognized.");
+}
+
+std::vector<std::string> ConsumeKafka::get_matching_headers(const rd_kafka_message_t* message, const std::string& header_name) const {
+  // Headers fetched this way are freed when rd_kafka_message_destroy is called
+  // Detaching them using rd_kafka_message_detach_headers does not seem to work
+  rd_kafka_headers_t* headers_raw;
+  const rd_kafka_resp_err_t get_header_response = rd_kafka_message_headers(message, &headers_raw);
+  if (RD_KAFKA_RESP_ERR__NOENT == get_header_response) {
+    return {};
+  }
+  if (RD_KAFKA_RESP_ERR_NO_ERROR != get_header_response) {
+    logger_->log_error("Failed to fetch message headers: %d: %s", rd_kafka_last_error(), rd_kafka_err2str(rd_kafka_last_error()));
+  }
+  std::vector<std::string> matching_headers;
+  for (std::size_t header_idx = 0;; ++header_idx) {
+    const char* value;  // Not to be freed
+    std::size_t size;
+    if (RD_KAFKA_RESP_ERR_NO_ERROR != rd_kafka_header_get(headers_raw, header_idx, header_name.c_str(), (const void**)(&value), &size)) {
+      break;
+    }
+    if (size < std::numeric_limits<int>::max()) {
+      logger_->log_debug("%.*s", static_cast<int>(size), value);
+    }
+    matching_headers.emplace_back(value, size);
+  }
+  return matching_headers;
+}
+
+std::vector<std::pair<std::string, std::string>> ConsumeKafka::get_flowfile_attributes_from_message_header(const rd_kafka_message_t* message) const {
+  std::vector<std::pair<std::string, std::string>> attributes_from_headers;
+  for (const std::string& header_name : headers_to_add_as_attributes_) {
+    const std::vector<std::string> matching_headers = get_matching_headers(message, header_name);
+    if (matching_headers.size()) {
+      attributes_from_headers.emplace_back(header_name, utils::get_encoded_string(resolve_duplicate_headers(matching_headers), message_header_encoding_attr_to_enum()));
+    }
+  }
+  return attributes_from_headers;
+}
+
+class WriteCallback : public OutputStreamCallback {
+ public:
+  WriteCallback(char *data, uint64_t size) :
+      data_(reinterpret_cast<uint8_t*>(data)),
+      dataSize_(size) {}
+  uint8_t* data_;
+  uint64_t dataSize_;
+  int64_t process(const std::shared_ptr<io::BaseStream>& stream) {
+    int64_t ret = 0;
+    if (data_ && dataSize_ > 0)
+      ret = stream->write(data_, dataSize_);
```

Comment:
It's quite unfortunate, but the size parameter of `BaseStream::write()` is an `int`, so this will cause a compiler warning.

At some point, we should probably bite the bullet and rewrite `BaseStream` to use `size_t` or `uint64_t` for stream lengths consistently, but for now, you can either change the type of `dataSize_` to `int` or add a `gsl::narrow` to silence the warning and prevent accidental (though very unlikely) truncation.

Reply:
Yep, during the implementation of ConsumeKafka, this stream writer suddenly changed and I thought just matching the types is good enough. It really makes no sense to check the sign of an `uint64_t`. Updated to use `gsl::narrow`.